extensions/lifecycle/bucketProcessor/LifecycleBucketProcessor.js

```diff
@@ -304,28 +305,92 @@ class LifecycleBucketProcessor {
-            } else {
-                task = new LifecycleTaskV2(this);
-            }
+            return this._createLifecycleTask(backbeatMetadataProxy, params, (err, task) => {
+                if (err) {
+                    this._log.error('failed to create lifecycle bucket task', {
+                        bucket,
+                        owner,
+                        error: err,
+                    });
+                    return cb(err);
+                }
 
-            this._log.info('scheduling new task for bucket lifecycle', {
-                method: 'LifecycleBucketProcessor._processBucketEntry',
-                bucket,
-                owner,
-                details: result.details,
-                taskName: task.constructor.name,
+                this._log.info('scheduling new task for bucket lifecycle', {
+                    method: 'LifecycleBucketProcessor._processBucketEntry',
+                    bucket,
+                    owner,
+                    details: result.details,
+                    taskName: task.constructor.name,
+                });
+                return this._internalTaskScheduler.push({
+                    task,
+                    rules: config.Rules,
+                    value: result,
+                    s3target: s3,
+                    backbeatMetadataProxy,
+                }, cb);
             });
-            return this._internalTaskScheduler.push({
-                task,
-                rules: config.Rules,
-                value: result,
-                s3target: s3,
-                backbeatMetadataProxy,
-            }, cb);
+        });
+
+    }
+
+    _createLifecycleTask(backbeatMetadataProxy, params, cb) {
+        let isV2 = false;
+        let task;
+
+        async.series([
+            next => backbeatMetadataProxy.getBucketIndexes(
+                params.bucket,
+                this._log,
+                (err, bucketIndexes) => {
+                    if (err) {
+                        this._log.warn('unable tparamso retrieve indexing info; using LifecycleTaskV1', {
+                            bucket: params.Bucket,
+                            error: err,
+                        });
+
+                        task = new LifecycleTask(this);
+                        return next(null);
+                    }
+
+                    isV2 = indexesForFeature.lifecycle.v2.every(
+                        idx => bucketIndexes.some(bidx => bidx.name === idx.name)
+                    );
+
+                    if (this._lcConfig.bucketProcessor.forceLegacyListing || !isV2) {
+                        task = new LifecycleTask(this);
+                        return next(null);
+                    }
+
+                    task = new LifecycleTaskV2(this);
+                    return next(null);
+
+                }),
+            next => {
+                if (!isV2) {
+                    return backbeatMetadataProxy.putBucketIndexes(
```

Comment:
Is this serialized (through the kafka message key/partition or otherwise)? Not sure what happens if a bunch of these requests are sent at once.